### PR TITLE
Boxplot fix median line extending past box boundaries #19409

### LINKED ISSUE #19409: Boxplot: Median line too long after changing linewidth
When changing the `linewidth` property of the median line in boxplots, the line will extend beyond the box width.

For example,
```python
import matplotlib.pyplot as plt

data = [[1, 2, 3, 4, 5]]
fig, ax = plt.subplots()
ax.boxplot(data,
           medianprops={"linewidth": 8})
```
gives (on Matplotlib 3.2.2):
[](https://user-images.githubusercontent.com/26203394/106387844-8a089680-63d3-11eb-944b-e22ead07b4ea.jpg)

By simply chaning the `solid_capstyle` propery to `"butt"` in
```python
ax.boxplot(data,
           medianprops={"linewidth": 8,
                        "solid_capstyle": "butt"})
```
I was able to improve this a bit (I stole that trick from [here](https://stackoverflow.com/a/10297860)):
[](https://user-images.githubusercontent.com/26203394/106387859-a0aeed80-63d3-11eb-8948-5ffac5306052.jpg)

Ideally, one could now reduce the median line length such that it no longer overlaps with the black bounding line of the box.
Since the box line width is only accessible in "points" units, it would need to be transformed to data units.
I was unable to get this to work, though.

Would it be possible/desirable to change the default setting for `solid_capstyle`?
Could this be handled in a different way?

### PR DESCRIPTION
**Pull request by kidkoder432**
## PR summary
Fixes #19409
The median line on a boxplot would extend past the box when the line width was increased. This PR aims to fix the visual annoyance by setting the solid line capstyle to be the same as all other capstyles. additionally, this change keeps the capstyle of the median lines consistent with other boxplot lines (i.e. mean lines). 

This is my second ever PR so be nice please :)

Other changes:
- Fixed a small typo in `bxp()` docstring
- Added a new test for the median line (from PR #23335)

## PR checklist

- [N/A] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [N/A] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-examples-and-tutorials)
- [N/A] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [x] Documentation complies with [general](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-docstrings) guidelines

### CODE REVIEW

**Comment by melissawm** [comment]
Hi @kidkoder432 - while we wait for a proper review, would you mind taking a look at the Linting failure here? https://github.com/matplotlib/matplotlib/actions/runs/5780696475/job/15664707497?pr=26462

It looks like you are missing a blank line in your test.

**Review by rcomer — COMMENTED** [review]
Thank you for your work on this @kidkoder432, but I do not think this is currently the right change.

**Comment by kidkoder432** [comment]
Oh okay, thanks! (This is my first contribution to matplotlib so I
appreciate the feedback :))

File: lib/matplotlib/axes/_axes.py
Comment (kidkoder432) [reply lib/matplotlib/axes/_axes.py:4098]: thanks :)

File: lib/matplotlib/lines.py
Comment (kidkoder432) [reply lib/matplotlib/lines.py]: its fixed now

**Comment by rcomer** [comment]
This change has caused some test failures.  @kidkoder432 are you confident to investigate those yourself?  You can see the full output from the tests by clicking the "Details" link by one of the checks at the bottom of this page.  You can also run the tests yourself locally:
https://matplotlib.org/devdocs/devel/testing.html

**Comment by kidkoder432** [comment]
> This change has caused some test failures. @kidkoder432 are you confident to investigate those yourself? You can see the full output from the tests by clicking the "Details" link by one of the checks at the bottom of this page. You can also run the tests yourself locally: https://matplotlib.org/devdocs/devel/testing.html

Sure, I'll take a look and let you know any concerns.

**Comment by kidkoder432** [comment]
Okay, so I looked at the failing tests and noticed that the expected images were all using a capstyle of `projecting`. However, because I modified the code to use a capstyle of `butt`. the original tests for median lines were failing because they were expecting a `projecting` capstyle. However, this capstyle is the reason why the median line was extending past the box. According to the [docs](https://matplotlib.org/stable/api/_enums_api.html#matplotlib._enums.CapStyle):

> `'projecting'`
the line is squared off as in `butt`, but the filled-in area extends beyond the endpoint a distance of `linewidth/2`.

I'm not sure how to move forward with this.

**Comment by rcomer** [comment]
OK, if the user specifies a capstyle then we should assume they are happy with their choice and leave it.  We should only change it to “butt” if the user didn’t make a choice.  I think the dictionary `setdefault` method might be useful here.

**Comment by kidkoder432** [comment]
> OK, if the user specifies a capstyle then we should assume they are happy with their choice and leave it. We should only set it to “butt” if the user didn’t make a choice. I think the dictionary `setdefault` method might be useful here.

Good idea. 
Those tests don't seem to define a capstyle specifically though, so they still fail...

**Comment by rcomer** [comment]
Ah OK, I thought you meant that some tests were specifying “projecting” as the capstyle.  Instead they are just leaving it to default and this change switches the default from “projecting” to “butt”.  So the tests are failing because of a genuine (desired) change in the output, which means you will need to replace the reference images.  We have the [triage_tests](https://github.com/matplotlib/matplotlib/blob/main/tools/triage_tests.py) tool to help with that.

**Comment by rcomer** [comment]
Thanks for the update @kidkoder432.  It looks like you still have a couple of image test failures for pdf files.  Do you see those failures locally?  If not you might be missing a dependency - most likely ghostscript.
https://matplotlib.org/devdocs/devel/dependencies.html#dependencies-for-testing-matplotlib

**Comment by kidkoder432** [comment]
i don't see them locally; i probably do need to install ghostscript.

How would I resolve the PDF test failures?

**Comment by rcomer** [comment]
I think the `triage_tests` script should handle pdf as well as png.

**Comment by kidkoder432** [comment]
Got it, thank you!

**Comment by ksunden** [comment]
Matplotlib only interacts with ghostscript via subprocess calls. Can you call `gs` as a CLI program (e.g. `gs --help`) (on windows it may also be called `gswin64c`, not really sure what is fully going on there...)

If it is not in your PATH, it will not be found, so make sure it is installed in a place that it can be executed.

**Comment by kidkoder432** [comment]
I use VSCode for development. If I run gswin64 from a regular powershell
terminal, it launches a command window to type gs commands with (gs is in
system PATH). However, this command does not work in VSCode, and pdf tests
are skipped by pytest.

How do I resolve this so i can continue development? Thanks.

**Comment by kidkoder432** [comment]
Never mind, i got it to work.

**Comment by kidkoder432** [comment]
Hi, for some reason it says that CodeCov tests failed in the workflow logs. However, when I visit the website, its says that codecov tests all passed. Can someone please take a look at this?

File: lib/matplotlib/axes/_axes.py
Comment (QuLogic) [inline lib/matplotlib/axes/_axes.py:4153]: Shouldn't use `setdefault` as that will modify the user's input dictionary. Something like this should avoid that:
```suggestion
        # Clamp median line to edge of box by default.
        medianprops = {
            "solid_capstyle": "butt",
            **(medianprops or {}),
        }
```

Comment (rcomer) [reply lib/matplotlib/axes/_axes.py:4153]: Ooops, `setdefault` was my suggestion.  Of course you’re right.

**Comment by rcomer** [comment]
I wouldn’t worry about that codecov failure.  We currently use codecov [for information only](https://matplotlib.org/devdocs/devel/coding_guide.html#automated-tests).  In this case, codecov says that 100% of your change is covered, which I think is the important thing.  The coverage reductions are all “indirect changes” - I haven’t really got my head around how those are caused myself.

File: .gitignore
Comment (rcomer) [inline .gitignore:18]: Could you reinstate this blank line?

File: lib/matplotlib/axes/_axes.py
Comment (kidkoder432) [reply lib/matplotlib/axes/_axes.py]: Just for my understanding, why exactly shouldn't we use setdefault? If the key exists in the dictionary, then setdefault won't make any changes. Am I understanding this correctly?

Comment (rcomer) [reply lib/matplotlib/axes/_axes.py]: Good question.

Suppose a user has a dictionary of properties that they re-use for multiple different plot types in their script.  If that dictionary gets added to when they pass it to `boxplot`, then it could affect other plots that are created further down the script.  That would be unexpected (and therefore undesirable) behaviour.

Does that make sense?  You are right that `setdefault` won’t change the dictionary if it already has that key.

**Comment by kidkoder432** [comment]
Yes, it makes sense. Thank you for the explanation!

**Review by rcomer — APPROVED** [review]
This looks good to me now.  For code changes, we require approvals from two reviewers so this can’t be merged just yet.  Thank you for your work on it @kidkoder432!

File: lib/matplotlib/axes/_axes.py
Comment (timhoffm) [inline lib/matplotlib/axes/_axes.py:4152]: I think we should unversally "solid_capstyle" and "dashed_capstyle" to "butt" for both `medianprops` and `meanprops`. We want the lines to end within the box for both independent of the global capstyle settings.

**Review by timhoffm — COMMENTED** [review]
Question: Does anybody understand why the box borders are sometimes shifted slightly?

Comment: The original issue also discussed bringing the line below the box using zorder. This is not discussed in this PR, While conceptually desireable, this is not so easy to fix: The box can be filled and then face and edge are drawn with the same zorder. This means, one cannot draw mean/median lines below the box edge but above the box face (at least not without complicating the draw logic).
I'm therefore inclinded to accept the PR without addressing zorder, but would like to have at least one second thought on this.

File: lib/matplotlib/axes/_axes.py
Comment (kidkoder432) [reply lib/matplotlib/axes/_axes.py:4152]: will do

**Comment by kidkoder432** [comment]
> Question: Does anybody understand why the box borders are sometimes shifted slightly?
> 
> Comment: The original issue also discussed bringing the line below the box using zorder. This is not discussed in this PR, While conceptually desireable, this is not so easy to fix: The box can be filled and then face and edge are drawn with the same zorder. This means, one cannot draw mean/median lines below the box edge but above the box face (at least not without complicating the draw logic).
> I'm therefore inclinded to accept the PR without addressing zorder, but would like to have at least one second thought on this.

Not sure why either... From what you explained about the draw logic it would likely be much easier to stay with the current implementation.

**Comment by rcomer** [comment]
> Question: Does anybody understand why the box borders are sometimes shifted slightly?

Do you have an example?  When I looked at the test images that have changed here, all the diffs seemed to just show the median line getting shorter.

> I'm therefore inclinded to accept the PR without addressing zorder, but would like to have at least one second thought on this.

I came to same conclusion based on @oscargus' comment on the previous PR: https://github.com/matplotlib/matplotlib/pull/23335#discussion_r905899991

**Comment by timhoffm** [comment]
> Do you have an example? When I looked at the test images that have changed here, all the diffs seemed to just show the median line getting shorter.

Can't find anymore. Either I was wrong or this was somehow fixed through the recent updates.

**Review by timhoffm — APPROVED** [review]

**Comment by timhoffm** [comment]
Since this is a bugfix, I recon that it can still go into 3.8.0 (on the basis that we do not have release candidates for any bugfix releases, there's no difference between deferring to 3.8.1 and pushing this into 3.8.0 even after 3.8.0rc1).

**Comment by timhoffm** [comment]
Thanks @kidkoder432, and congratulations on your first contrbution to matplotlib! We'd happy to see you back!

**Comment by kidkoder432** [comment]
You're welcome! Thank you for your support!
I'll definitely be back to contribute :)

**Pull request merged**
